File: arrays_and_hashing/06_encode_decode.cpp
#include <vector>
#include <string>
#include <cassert>
#include <cstdio>
#include <iostream>

using std::vector, std::string;

string encode(const vector<string>& strs) {
	string result;
	for (const auto& s : strs) {
		result += std::to_string(s.size()) + ";" + s;
	}
	return result;
}

vector<string> decode(const string s) {
	vector<string> result;

	int i = 0;
	while (i < s.size()) {
		int j = i;
		while (s[j] != ';') {
			j++;
		}
		int length = std::stoi(s.substr(i, j - i));
		result.push_back(s.substr(j+1, length));
		i = j+1+length;
	}
	return result;
}

void test_encode_decode()
{
	vector<vector<string>> str_cases = {
		vector<string>{"test", "test", "test"},
		vector<string>{"1num", "num1", ";;", ";1;3"},
		vector<string>{"test", "longassstring"},
		vector<string>{"test"},
	};
	for (const auto& str_case : str_cases) {
		assert(decode(encode(str_case)) == str_case);
	}

	printf("All tests passed for encode and decode.\n");
}

int main()
{
	test_encode_decode();
	return 0;
}

File: arrays_and_hashing/06_encode_decode.py
def test_encode_decode():
    cases = [
        ["we","say",":","yes","!@#$%^&*()"],
        ["a", "b", "c"],
        ["a"],
        ["test", "test", "test"],
        ["python", "java", "c++"],
        ["", "test", "test"],
        ["", "", ""],
        [""],
        ["1num", "num1", ";;", ";1;3"]
    ]
    for string_l in cases:
        assert decode(encode(string_l)) == string_l, string_l


def encode(strs: list[str]) -> str:
    res = []
    for s in strs:
        res.append(str(len(s)) + ';' + s)
    return "".join(res)


def decode(s: str) -> list[str]:
    res = []
    i = 0
    while i < len(s):
        digits = []
        j = 0
        while (digit := s[i + j]) != ';':
            digits.append(digit)
            j += 1
        assert s[i + len(digits)] == ';'
        string_length = int("".join(digits))
        current_string_position = len(digits) + 1 + i
        next_string_position = current_string_position + string_length
        string_content = s[current_string_position:next_string_position]
        res.append(string_content)
        i = next_string_position
    return res


if __name__ == "__main__":
    test_encode_decode()
